File: pkg/logs/redact/redact.go
// Package redact provides log redaction functionality to prevent secret leakage in CI logs.
package redact

import (
	"fmt"
	"math"
	"os"
	"regexp"
	"strings"
)

// Mode represents the redaction mode.
type Mode string

const (
	// ModeOff disables redaction.
	ModeOff Mode = "off"
	// ModeBasic enables basic redaction (default in CI).
	ModeBasic Mode = "basic"
	// ModeAggressive enables aggressive redaction with high-entropy detection.
	ModeAggressive Mode = "aggressive"

	// minEntropyCandidateLen is the minimum token length considered for entropy-based redaction.
	minEntropyCandidateLen = 20
)

// DefaultMode is the default redaction mode when running in CI.
// It's "basic" when CI is detected, "off" otherwise.
var DefaultMode = func() Mode {
	if isCI() {
		return ModeBasic
	}
	return ModeOff
}()

// Redactor handles log redaction.
type Redactor struct {
	mode           Mode
	customKeys     []string
	replacement    string
}

// Config holds configuration for a Redactor.
type Config struct {
	Mode        Mode   // Redaction mode: off, basic, aggressive
	CustomKeys  string // Comma-separated list of custom key patterns (e.g., "MY_API_KEY,SECRET_TOKEN")
	Replacement string // Replacement string (default: "***REDACTED***")
}

// New creates a new Redactor with the given configuration.
func New(cfg Config) *Redactor {
	mode := cfg.Mode
	if mode == "" {
		mode = DefaultMode
	}

	replacement := cfg.Replacement
	if replacement == "" {
		replacement = "***REDACTED***"
	}

	customKeys := []string{}
	if cfg.CustomKeys != "" {
		for _, key := range strings.Split(cfg.CustomKeys, ",") {
			key = strings.TrimSpace(key)
			if key != "" {
				customKeys = append(customKeys, key)
			}
		}
	}

	return &Redactor{
		mode:        mode,
		customKeys:  customKeys,
		replacement: replacement,
	}
}

// RedactFile reads a file, redacts sensitive content, and writes it back.
func (r *Redactor) RedactFile(path string) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("failed to read file %s: %w", path, err)
	}

	redacted, err := r.Redact(data)
	if err != nil {
		return fmt.Errorf("failed to redact content from %s: %w", path, err)
	}

	if err := os.WriteFile(path, redacted, 0644); err != nil {
		return fmt.Errorf("failed to write redacted content to %s: %w", path, err)
	}

	return nil
}

// Redact redacts sensitive content from the input data.
func (r *Redactor) Redact(data []byte) ([]byte, error) {
	if r.mode == ModeOff {
		return data, nil
	}

	content := string(data)
	var err error

	// Apply all redaction passes
	content = r.redactPEMBlocks(content)
	content = r.redactEnvKeyValues(content)
	content = r.redactHTTPHeaderValues(content)
	content = r.redactURLQueryParams(content)

	if r.mode == ModeAggressive {
		content, err = r.redactHighEntropyStrings(content)
		if err != nil {
			return nil, fmt.Errorf("failed to redact high-entropy strings: %w", err)
		}
		content = r.redactKnownPrefixes(content)
	}

	return []byte(content), nil
}

// redactEnvKeyValues redacts environment variable key-value pairs.
func (r *Redactor) redactEnvKeyValues(content string) string {
	// Common secret patterns in env keys
	patterns := []string{
		`_TOKEN`, `_KEY`, `_SECRET`, `_PASSWORD`,
		`_API_KEY`, `_AUTH_TOKEN`, `_AUTHORIZATION`,
	}

	// Add custom key patterns
	for _, key := range r.customKeys {
		patterns = append(patterns, key)
	}

	// Also check for exact matches (no underscore)
	exactPatterns := []string{
		`API_KEY`, `AUTH_TOKEN`, `PASSWORD`, `APIKEY`, `SECRET`,
	}

	// Pattern for: KEY=VALUE or KEY="VALUE" or KEY='VALUE'
	// Match env-like assignments in logs
	for _, pattern := range patterns {
		// Convert glob pattern to regex (e.g., *_TOKEN -> \w+_TOKEN)
		regexPattern := strings.ReplaceAll(regexp.QuoteMeta(pattern), "\\_", "\\w*_")
		regexPattern = `(\w+` + regexPattern + `)\s*=\s*[\'"]?([^\'"\s\n]+)[\'"]?`

		re := regexp.MustCompile(regexPattern)
		content = re.ReplaceAllString(content, fmt.Sprintf("$1=%s", r.replacement))
	}

	// Handle exact matches
	for _, pattern := range exactPatterns {
		re := regexp.MustCompile(`(`+pattern+`)\s*=\s*[\'"]?([^\'"\s\n]+)[\'"]?`)
		content = re.ReplaceAllString(content, fmt.Sprintf("$1=%s", r.replacement))
	}

	return content
}

// redactHTTPHeaderValues redacts HTTP header values.
func (r *Redactor) redactHTTPHeaderValues(content string) string {
	// Sensitive HTTP headers
	headers := []string{
		"Authorization", "X-API-Key", "X-Auth-Token",
		"X-GitHub-Token", "X-Api-Key", "Authentication",
		"Cookie", "Set-Cookie", "Proxy-Authorization",
	}

	for _, header := range headers {
		// Match: Header: value (case insensitive) while preserving the original header casing in output.
		re := regexp.MustCompile(`(?i)(^|\n)\s*(` + regexp.QuoteMeta(header) + `)\s*:\s*[^\n\r]+`)
		content = re.ReplaceAllString(content, fmt.Sprintf("$1$2: %s", r.replacement))
	}

	return content
}

// redactURLQueryParams redacts sensitive URL query parameters.
func (r *Redactor) redactURLQueryParams(content string) string {
	// Sensitive query param names
	params := []string{
		"token", "key", "secret", "password", "api_key",
		"access_token", "refresh_token", "auth_token",
		"apikey", "authorization",
	}

	for _, param := range params {
		// Match: ?param=value or &param=value or &param=value&
		// Handle various URL formats
		re := regexp.MustCompile(`([?&])` + regexp.QuoteMeta(param) + `=[^&\s#\'"]+`)
		content = re.ReplaceAllString(content, fmt.Sprintf("$1%s=%s", param, r.replacement))
	}

	return content
}

// redactPEMBlocks redacts PEM-encoded private keys and certificates.
func (r *Redactor) redactPEMBlocks(content string) string {
	// Match PEM blocks: -----BEGIN ...----- to -----END ...-----
	// The label can contain spaces, letters, numbers, and common symbols
	re := regexp.MustCompile(`-----BEGIN [A-Za-z0-9+/\s-]+-----[\s\S]*?-----END [A-Za-z0-9+/\s-]+-----`)
	return re.ReplaceAllString(content, fmt.Sprintf("-----BEGIN REDACTED-----\n%s\n-----END REDACTED-----", r.replacement))
}

// redactKnownPrefixes redacts strings with known secret prefixes.
func (r *Redactor) redactKnownPrefixes(content string) string {
	// Known API key/token prefixes with their typical lengths
	patterns := []struct {
		prefix   string
		minLen   int
		maxLen   int
		pattern  string
	}{
		// GitHub tokens
		{prefix: "ghp_", minLen: 36, maxLen: 40, pattern: "ghp_[A-Za-z0-9_]{32,36}"},
		{prefix: "gho_", minLen: 36, maxLen: 40, pattern: "gho_[A-Za-z0-9_]{32,36}"},
		{prefix: "ghu_", minLen: 36, maxLen: 40, pattern: "ghu_[A-Za-z0-9_]{32,36}"},
		{prefix: "ghs_", minLen: 36, maxLen: 40, pattern: "ghs_[A-Za-z0-9_]{32,36}"},
		{prefix: "ghr_", minLen: 36, maxLen: 40, pattern: "ghr_[A-Za-z0-9_]{32,36}"},
		// Stripe
		{prefix: "sk_live_", minLen: 40, maxLen: 50, pattern: "sk_live_[A-Za-z0-9_]{32,40}"},
		{prefix: "sk_test_", minLen: 40, maxLen: 50, pattern: "sk_test_[A-Za-z0-9_]{32,40}"},
		{prefix: "sk-", minLen: 30, maxLen: 50, pattern: "sk-[A-Za-z0-9_]{26,46}"},
		// Anthropic
		{prefix: "anthropic_", minLen: 40, maxLen: 60, pattern: "anthropic_[A-Za-z0-9_]{30,50}"},
		// HuggingFace
		{prefix: "hf_", minLen: 30, maxLen: 50, pattern: "hf_[A-Za-z0-9_]{26,46}"},
		// AWS
		{prefix: "AKIA", minLen: 20, maxLen: 20, pattern: "AKIA[A-Z0-9]{16}"},
		// Slack
		{prefix: "xoxb-", minLen: 30, maxLen: 50, pattern: "xoxb-[A-Za-z0-9\\-]{26,46}"},
		{prefix: "xoxp-", minLen: 30, maxLen: 50, pattern: "xoxp-[A-Za-z0-9\\-]{26,46}"},
		// Google OAuth
		{prefix: "ya29.", minLen: 50, maxLen: 200, pattern: "ya29\\.[A-Za-z0-9_\\-]{46,196}"},
	}

	result := content
	for _, p := range patterns {
		re := regexp.MustCompile(p.pattern)
		result = re.ReplaceAllString(result, p.prefix+r.replacement)
	}

	return result
}

// redactHighEntropyStrings redacts strings that appear to be high-entropy secrets.
// This is a best-effort heuristic and may have false positives.
func (r *Redactor) redactHighEntropyStrings(content string) (string, error) {
	// Look for potential secrets (strings that look like base64, hex, or random strings)
	// Look for:
	// - Base64-like strings (20+ chars)
	// - Mixed alphanumeric with symbols (16+ chars)

	// Regex to find potential secrets
	potentialSecretRe := regexp.MustCompile(
		fmt.Sprintf(`\b[A-Za-z0-9_\-\.]{%d,}\b`, minEntropyCandidateLen),
	)

	lines := strings.Split(content, "\n")
	var redactedLines []string

	for _, line := range lines {
		matches := potentialSecretRe.FindAllString(line, -1)

		for _, match := range matches {
			// Skip if it's a common word or path
			if isLikelyFalsePositive(match) {
				continue
			}

			// Check for high entropy
			if r.isHighEntropy(match) {
				line = strings.ReplaceAll(line, match, r.replacement)
			}
		}

		redactedLines = append(redactedLines, line)
	}

	return strings.Join(redactedLines, "\n"), nil
}

// isHighEntropy calculates Shannon entropy of a string to determine if it looks like a secret.
func (r *Redactor) isHighEntropy(s string) bool {
	if len(s) < minEntropyCandidateLen {
		return false
	}

	// Calculate character frequency
	freq := make(map[rune]float64)
	for _, ch := range s {
		freq[ch]++
	}

	// Calculate Shannon entropy
	entropy := 0.0
	for _, count := range freq {
		p := count / float64(len(s))
		if p > 0 {
			entropy -= p * math.Log2(p)
		}
	}

	// Threshold: entropy > 4.0 suggests high randomness (like a secret)
	// Most natural language has entropy < 3.5
	return entropy > 4.0
}

// isLikelyFalsePositive checks if a string is likely a false positive (not a secret).
func isLikelyFalsePositive(s string) bool {
	// Check if it's a path
	if strings.Contains(s, "/") || strings.Contains(s, "\\") {
		return true
	}

	// Check if it's a URL
	if strings.HasPrefix(s, "http://") || strings.HasPrefix(s, "https://") {
		return true
	}

	// Check if it's all lowercase (likely a word)
	if s == strings.ToLower(s) && len(s) < 30 {
		return true
	}

	// Check if it's all uppercase (likely an acronym)
	if s == strings.ToUpper(s) && len(s) < 20 {
		return true
	}

	// Check if it looks like base64 but contains mostly lowercase letters
	lowerCount := 0
	for _, ch := range s {
		if ch >= 'a' && ch <= 'z' {
			lowerCount++
		}
	}
	if float64(lowerCount)/float64(len(s)) > 0.7 {
		return true
	}

	return false
}

// isCI detects if running in a CI environment.
func isCI() bool {
	// Check common CI environment variables
	ciEnvs := []struct {
		name  string
		truthy string
	}{
		{"CI", ""},
		{"GITHUB_ACTIONS", "true"},
		{"GITLAB_CI", "true"},
		{"TRAVIS", "true"},
		{"CIRCLECI", "true"},
		{"JENKINS_URL", ""},
		{"BUILDKITE", "true"},
		{"BITBUCKET_BUILD_NUMBER", ""},
		{"AZURE_PIPELINES", "true"},
	}

	for _, env := range ciEnvs {
		if _, exists := os.LookupEnv(env.name); exists {
			// For CI env, check if it's set to a truthy value
			val := os.Getenv(env.name)
			if env.truthy == "" || val == env.truthy || val == "1" || val == "true" {
				return true
			}
		}
	}

	return false
}

// RedactFromEnv creates a Redactor from environment variables.
// Uses HOLON_LOG_REDACT for mode and HOLON_LOG_REDACT_KEYS for custom keys.
func RedactFromEnv() *Redactor {
	modeStr := os.Getenv("HOLON_LOG_REDACT")
	mode := Mode(modeStr)
	if modeStr == "" {
		mode = DefaultMode
	}

	// Validate mode
	switch mode {
	case ModeOff, ModeBasic, ModeAggressive:
		// Valid modes
	default:
		mode = DefaultMode
	}

	customKeys := os.Getenv("HOLON_LOG_REDACT_KEYS")
	replacement := os.Getenv("HOLON_LOG_REDACT_REPLACEMENT")

	return New(Config{
		Mode:        mode,
		CustomKeys:  customKeys,
		Replacement: replacement,
	})
}

File: agents/claude/src/redact.ts
/**
 * Redaction utility for tool parameters in logs.
 *
 * This module provides functions to redact sensitive information from tool
 * parameters before they are displayed in logs, preventing secret leakage.
 */

/**
 * Sensitive field patterns that should be redacted from tool parameters.
 * These match common names for sensitive fields in tool inputs.
 */
const SENSITIVE_PATTERNS = [
  // Direct keys
  /^token$/i,
  /^access_token$/i,
  /^api_key$/i,
  /^apikey$/i,
  /^secret$/i,
  /^password$/i,
  /^passwd$/i,
  /^authorization$/i,
  /^auth$/i,
  /^bearer$/i,
  /^credential$/i,
  /^cookie$/i,
  /^session(id)?$/i,
  /^session_token$/i,
  /^refresh_token$/i,
  /^auth_token$/i,
  /^private_key$/i,
  /^passphrase$/i,
  // Header patterns (e.g., headers.authorization)
  /^headers\.(authorization|cookie|set-cookie)$/i,
  /^headers\.[xX]-[aA]pi-[kK]ey$/i,
  /^headers\.[xX]-[aA]uth-[tT]oken$/i,
];

/**
 * Patterns for detecting high-entropy strings that might be secrets.
 */
const HIGH_ENTROPY_PREFIXES = [
  "ghp_", "gho_", "ghu_", "ghs_", "ghr_", // GitHub
  "sk_live_", "sk_test_", "sk-", // Stripe
  "anthropic_", // Anthropic
  "hf_", // HuggingFace
  "AKIA", // AWS
  "xoxb-", "xoxp-", // Slack
  "ya29.", // Google OAuth
];

/**
 * Replaces sensitive values with a placeholder string.
 * @param value - The value to potentially redact
 * @param key - The key name (used for pattern matching)
 * @param replacement - The replacement string (default: "***")
 * @returns The original value or replacement string
 */
export function redactValue(value: unknown, key: string, replacement = "***"): unknown {
  // Only redact strings
  if (typeof value !== "string") {
    return value;
  }

  // Check if the key matches a sensitive pattern
  for (const pattern of SENSITIVE_PATTERNS) {
    if (pattern.test(key)) {
      return replacement;
    }
  }

  // Check for known API key/token prefixes
  const lowerValue = value.toLowerCase();
  for (const prefix of HIGH_ENTROPY_PREFIXES) {
    if (value.startsWith(prefix) || lowerValue.startsWith(prefix.toLowerCase())) {
      return replacement;
    }
  }

  // Check for high-entropy strings (heuristic)
  if (looksLikeHighEntropySecret(value)) {
    return replacement;
  }

  return value;
}

/**
 * Redacts sensitive values from an object recursively.
 * @param obj - The object to sanitize
 * @param replacement - The replacement string (default: "***")
 * @returns A new object with sensitive values redacted
 */
export function sanitizeToolInput<T extends Record<string, unknown>>(
  obj: T | undefined,
  replacement = "***"
): Record<string, unknown> | undefined {
  if (!obj || typeof obj !== "object") {
    return undefined;
  }

  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (value === null || value === undefined) {
      continue;
    }

    // Recursively handle nested objects
    if (typeof value === "object" && !Array.isArray(value)) {
      const sanitized = sanitizeToolInput(value as Record<string, unknown>, replacement);
      if (sanitized && Object.keys(sanitized).length > 0) {
        result[key] = sanitized;
      }
      continue;
    }

    // Handle arrays (recursively sanitize objects inside)
    if (Array.isArray(value)) {
      const sanitizedArray: unknown[] = [];
      for (const item of value) {
        if (typeof item === "object" && item !== null) {
          const sanitized = sanitizeToolInput(item as Record<string, unknown>, replacement);
          if (sanitized) {
            sanitizedArray.push(sanitized);
          }
        } else {
          sanitizedArray.push(redactValue(item, key, replacement));
        }
      }
      if (sanitizedArray.length > 0) {
        result[key] = sanitizedArray;
      }
      continue;
    }

    // Redact sensitive values
    const redacted = redactValue(value, key, replacement);
    if (redacted !== null && redacted !== undefined) {
      result[key] = redacted;
    }
  }

  return result;
}

/**
 * Formats tool input parameters for logging in a compact, readable way.
 * @param toolName - Name of the tool being executed
 * @param toolInput - The input parameters to the tool
 * @returns A formatted string representation of the tool input
 */
export function formatToolInputForLog(toolName: string, toolInput: Record<string, unknown> | undefined): string {
  if (!toolInput || Object.keys(toolInput).length === 0) {
    return toolName;
  }

  // Sanitize the input first
  const sanitized = sanitizeToolInput(toolInput);

  // Format based on tool type
  const parts: string[] = [toolName];

  // Tool-specific formatting
  switch (toolName) {
    case "Bash": {
      const cmd = sanitized?.command;
      if (typeof cmd === "string") {
        // Show command, truncated if too long
        parts.push(formatValue("command", cmd, 100));
      }
      break;
    }
    case "Read":
    case "Write":
    case "Edit": {
      if (typeof sanitized?.file_path === "string") {
        parts.push(formatValue("file_path", sanitized.file_path, 80));
      }
      break;
    }
    case "Grep": {
      if (typeof sanitized?.pattern === "string") {
        parts.push(formatValue("pattern", sanitized.pattern, 40));
      }
      if (typeof sanitized?.path === "string") {
        parts.push(formatValue("path", sanitized.path, 60));
      }
      break;
    }
    case "Glob": {
      if (typeof sanitized?.pattern === "string") {
        parts.push(formatValue("pattern", sanitized.pattern, 60));
      }
      break;
    }
    case "HTTP": {
      if (typeof sanitized?.method === "string") {
        parts.push(sanitized.method.toUpperCase());
      }
      if (typeof sanitized?.url === "string") {
        parts.push(formatValue("url", sanitized.url, 80));
      }
      break;
    }
    case "Skill": {
      if (typeof sanitized?.skill === "string") {
        parts.push(formatValue("skill", sanitized.skill, 40));
      }
      if (typeof sanitized?.args === "string") {
        parts.push(formatValue("args", sanitized.args, 60));
      }
      break;
    }
    default: {
      // Generic formatting: show up to 2 key parameters
      if (sanitized) {
        let count = 0;
        const priorityKeys = ["file_path", "pattern", "command", "query", "description", "prompt", "url"];
        for (const key of priorityKeys) {
          if (key in sanitized && count < 2) {
            const value = (sanitized as Record<string, unknown>)[key];
            if (value !== undefined) {
              parts.push(formatValue(key, String(value), 60));
              count++;
            }
          }
        }
      }
    }
  }

  return parts.join(" ");
}

/**
 * Formats a key-value pair for logging.
 * @param key - The parameter name
 * @param value - The parameter value
 * @param maxLength - Maximum length for the value (will be truncated if longer)
 * @returns A formatted string like 'key="value"'
 */
function formatValue(key: string, value: string, maxLength = 80): string {
  const truncated = value.length > maxLength ? value.slice(0, maxLength - 3) + "..." : value;
  // Escape quotes and backslashes
  const escaped = truncated.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
  return `${key}="${escaped}"`;
}

/**
 * Checks if a string looks like a high-entropy secret.
 * This is a heuristic to catch things like random API keys.
 */
function looksLikeHighEntropySecret(str: string): boolean {
  // Skip short strings
  if (str.length < 20) {
    return false;
  }

  // Skip things that look like paths or URLs
  if (str.includes("/") || str.includes("\\") || str.startsWith("http")) {
    return false;
  }

  // Calculate Shannon entropy (simplified)
  const freq: Record<string, number> = {};
  for (const ch of str) {
    freq[ch] = (freq[ch] || 0) + 1;
  }

  let entropy = 0;
  const len = str.length;
  for (const count of Object.values(freq)) {
    const p = count / len;
    if (p > 0) {
      entropy -= p * Math.log2(p);
    }
  }

  // High entropy threshold (random-looking strings have entropy > 4.0)
  return entropy > 4.0;
}

/**
 * Checks if a value should be completely omitted from logs (not just redacted).
 * This is for values that are too large or not useful to show.
 */
export function shouldOmitFromLogs(key: string, value: unknown): boolean {
  // Omit very long strings
  if (typeof value === "string" && value.length > 500) {
    return true;
  }

  // Omit large arrays or objects
  if (Array.isArray(value) && value.length > 10) {
    return true;
  }

  // Omit certain keys entirely
  const omitKeys = ["content", "text", "body", "data", "output"];
  return omitKeys.some((omit) => key.toLowerCase() === omit);
}
